src/include/network/routines.rb

```diff
@@ -513,138 +379,81 @@ def ReadHardware(hwtype)
         # ones (which may be smaller). #102945.
         one["num"] = num
 
-        case controller
-          # modem
-        when "modem"
-          one["device_name"] = card["dev_name"] || ""
-          one["drivers"] = card["drivers"] || []
-
-          speed = Ops.get_integer(resource, ["baud", 0, "speed"], 57_600)
-          # :-) have to check .probe and libhd if this confusion is
-          # really necessary. maybe a pppd bug too? #148893
-          speed = 57_600 if speed == 12_000_000
-
-          one["speed"] = speed
-          one["init1"] = Ops.get_string(resource, ["init_strings", 0, "init1"], "")
-          one["init2"] = Ops.get_string(resource, ["init_strings", 0, "init2"], "")
-          one["pppd_options"] = Ops.get_string(resource, ["pppd_option", 0, "option"], "")
-
-          # isdn card
-        when "isdn"
-          drivers = card["isdn"] || []
-          one["drivers"] = drivers
-          one["sel_drv"] = 0
-          one["bus"] = card["bus"] || ""
-          one["io"] = Ops.get_integer(resource, ["io", 0, "start"], 0)
-          one["irq"] = Ops.get_integer(resource, ["irq", 0, "irq"], 0)
-
-          # dsl card
-        when "dsl"
-          driver_info = Ops.get_map(card, ["dsl", 0], {})
-          translate_mode = { "capiadsl" => "capi-adsl", "pppoe" => "pppoe" }
-          m = driver_info["mode"] || ""
-          one["pppmode"] = translate_mode[m] || m
-
-          # treat the rest as a network card
+        # drivers:
+        # Although normally there is only one module
+        # (one=$[active:, module:, options:,...]), the generic
+        # situation is: one or more driver variants (exclusive),
+        # each having one or more modules (one[drivers])
+
+        # only drivers that are not marked as broken (#97540)
+        drivers = card["drivers"].select do |d|
+          # ignoring more modules per driver...
+          first_module = d.fetch("modules", []).fetch(0, []).fetch(0, []) # [module, options]
```

Comment:
For me the last fetch looks like bug as include? below looks like expect module name as it use split on broken_modules. so it never find [] in broken_modules. so it actually works, but it is confusing.